File: Data Structures & Algorithms/minimum-stack/submission-1.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.min_stack = []  # keeps track of the min at each level

    def push(self, val: int) -> None:
        self.stack.append(val)
        # If min_stack is empty or val is smaller/equal to current min, push it
        if not self.min_stack or val <= self.min_stack[-1]:
            self.min_stack.append(val)

    def pop(self) -> None:
        if self.stack:
            val = self.stack.pop()
            if val == self.min_stack[-1]:  # if we popped the min, remove it from min_stack too
                self.min_stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]

File: Data Structures & Algorithms/daily-temperatures/submission-1.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        stack<int> s;
        vector<int> res(temperatures.size(), 0);

        for (int i =0; i < temperatures.size(); i++){
            if (s.empty()){
                s.push(i);
            } else if ( temperatures[s.top()] < temperatures[i]){
                while (!s.empty() && temperatures[i] > temperatures[s.top()]){
                    int index = s.top();
                    res[index] = i - index;
                    s.pop();
                }
            }
            s.push(i);
        }
        return res;
    }
};
